### PR TITLE
Add table aws_waf_rate_based_rule. closes #288

### LINKED ISSUE #288: Add table aws_waf_rate_based_rule
**References**
https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.37.26/service/waf

### PR DESCRIPTION
**Pull request by Priyanka585464**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
SETUP: tests/aws_waf_rate_based_rule []

PRETEST: tests/aws_waf_rate_based_rule

TEST: tests/aws_waf_rate_based_rule
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.aws_partition.current: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_waf_rate_based_rule.named_test_resource: Creating...
aws_waf_rate_based_rule.named_test_resource: Creation complete after 7s [id=82adc900-333c-4748-8826-efeae8d78cc0]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

account_id = 986325076436
aws_partition = aws
resource_aka = arn:aws:waf::986325076436:ratebasedrule/82adc900-333c-4748-8826-efeae8d78cc0
resource_id = 82adc900-333c-4748-8826-efeae8d78cc0
resource_name = turbottest98262

Running SQL query: test-get-query.sql
[
  {
    "metric_name": "turbottest98262",
    "partition": "aws",
    "title": "turbottest98262"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:waf::986325076436:ratebasedrule/82adc900-333c-4748-8826-efeae8d78cc0"
    ],
    "metric_name": "turbottest98262",
    "tags": {
      "name": "turbottest98262"
    },
    "title": "turbottest98262"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "metric_name": "turbottest98262",
    "partition": "aws",
    "title": "turbottest98262"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "account_id": "986325076436",
    "akas": [
      "arn:aws:waf::986325076436:ratebasedrule/82adc900-333c-4748-8826-efeae8d78cc0"
    ],
    "title": "turbottest98262"
  }
]
✔ PASSED

POSTTEST: tests/aws_waf_rate_based_rule

TEARDOWN: tests/aws_waf_rate_based_rule

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```
### Basic info

```sql
select
  name,
  rule_id,
  metric_name
from
  aws_waf_rate_based_rule;
```
```
+-----------+--------------------------------------+-------------+
| name      | rule_id                              | metric_name |
+-----------+--------------------------------------+-------------+
| pikuturb  | e163b892-37ff-4690-b5d0-9a0eb4ff87da | pikuturb    |
| tfWAFRule | 65c01a91-fc38-4db7-ba19-23487d7dd12f | tfWAFRule   |
+-----------+--------------------------------------+-------------+
```

### List ratebasedrules whose predicates are based on negation

```sql
select
  name,
  rule_id,
  p ->> 'DataId' as data_id,
  p ->> 'Negated' as negated,
  p ->> 'Type' as type
from
  aws_waf_rate_based_rule,
  jsonb_array_elements(predicates) as p
where
  p ->> 'Negated' = 'True';
```
```
+------+---------+---------+---------+------+
| name | rule_id | data_id | negated | type |
+------+---------+---------+---------+------+
+------+---------+---------+---------+------+
```
(Note: Since here negated value is set to false)
```
+-----------+--------------------------------------+--------------------------------------+---------+---------+
| name      | rule_id                              | data_id                              | negated | type    |
+-----------+--------------------------------------+--------------------------------------+---------+---------+
| tfWAFRule | 65c01a91-fc38-4db7-ba19-23487d7dd12f | 4cab3814-ea7c-4237-ae65-f8403a2d179c | false   | IPMatch |
+-----------+--------------------------------------+--------------------------------------+---------+---------+
```

### CODE REVIEW

File: aws/table_aws_waf_rate_based_rule.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	"github.com/turbot/steampipe-plugin-sdk/grpc/proto"
+	"github.com/turbot/steampipe-plugin-sdk/plugin"
+	"github.com/turbot/steampipe-plugin-sdk/plugin/transform"
+)
+
+//// TABLE DEFINITION
+func tableAwsWafRateBasedRule(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "aws_waf_rate_based_rule",
+		Description: "AWS WAF RateBasedRule",
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:16]:
```suggestion
		Description: "AWS WAF Rate Based Rule",
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.SingleColumn("rule_id"),
+			ShouldIgnoreError: isNotFoundError([]string{"WAFNonexistentItemException", "ValidationException"}),
+			Hydrate:           getAwsWafRateBasedRules,
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:20]:
```suggestion
			Hydrate:           getAwsWafRateBasedRule,
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+			},
+			{
+				Name:        "tags_src",
+				Description: "A list of tags assigned to the Rule.",
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     listAwsWafRateBasedRuleTags,
+				Transform:   transform.FromField("TagInfoForResource.TagList"),
+			},
+
+			// // Standard columns for all tables
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:69]:
```suggestion
			// Standard columns for all tables
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+			params.NextMarker = response.NextMarker
+		} else {
+			pagesLeft = false
+		}
+	}
+	return nil, nil
+}
+
+////  HYDRATE FUNCTIONS
+func getAwsWafRateBasedRules(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:124]:
```suggestion
func getAwsWafRateBasedRule(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+	data, err := svc.GetRateBasedRule(params)
+	if err != nil {
+		logger.Debug("getAwsWafRateBasedRules", "ERROR", err)
+		return nil, err
+	}
+	return data.Rule, nil
+}
+
+//// TRANSFORM FUNCTIONS
+
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:152]: are listAwsWafRateBasedRuleTags and getAwsWafRateBasedRuleAkas transform functions?

File: aws/table_aws_waf_rate_based_rule.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	aka := "arn:" + commonColumnData.Partition + ":waf::" + commonColumnData.AccountId + ":ratebasedrule" + "/" + id
+	return []string{aka}, nil
+}
+
+func tagListToTurbotTags(ctx context.Context, d *transform.TransformData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("tagListToTurbotTags")
+	tagList := d.HydrateItem.(*waf.ListTagsForResourceOutput)
+	if tagList.TagInfoForResource.TagList == nil {
+		return nil, nil
+	}
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:203]:
```suggestion
	}
	
```

File: aws-test/tests/aws_waf_rate_based_rule/test-hydrate-query.sql [hunk 7]
```diff
@@ -0,0 +1,4 @@
+select metric_name, akas, title
```
Comment (Paulami30) [inline aws-test/tests/aws_waf_rate_based_rule/test-hydrate-query.sql:1]:
```suggestion
select metric_name, akas, title, tags
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+			},
+			{
+				Name:        "metric_name",
+				Description: "The name or description for the metrics for a RateBasedRule.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getAwsWafRateBasedRules,
+			},
+			{
+				Name:        "predicates",
+				Description: "The format of the file that contains the IPSet.",
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:44]:
```suggestion
				Description: "The Predicates object contains one Predicate element for each ByteMatchSet, IPSet or SqlInjectionMatchSet object that you want to include in a RateBasedRule.",
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getAwsWafRateBasedRules,
+				Transform:   transform.FromField("MatchPredicates"),
+			},
+			{
+				Name:        "rate_key",
+				Description: "The name or description for the metrics for a RateBasedRule.",
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:51]:
```suggestion
				Description: "The field that AWS WAF uses to determine if requests are likely arriving from single source and thus subject to rate monitoring.",
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getAwsWafRateBasedRules,
+			},
+			{
+				Name:        "rate_limit",
+				Description: "The name or description for the metrics for a RateBasedRule.",
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:57]:
```suggestion
				Description: "The maximum number of requests, which have an identical value in the field specified by the RateKey, allowed in a five-minute period.",
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+				Type:        proto.ColumnType_STRING,
+			},
+			{
+				Name:        "metric_name",
+				Description: "The name or description for the metrics for a RateBasedRule.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getAwsWafRateBasedRules,
+			},
+			{
+				Name:        "predicates",
```
Comment (Paulami30) [inline aws/table_aws_waf_rate_based_rule.go:43]: arrange as the standard column format

File: docs/tables/aws_waf_rate_based_rule.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+select
+  name,
+  rule_id,
+  metric_name
+from
+  aws_waf_rate_based_rule;
+```
+
+
+### List ratebasedrules whose predicates are based on negation
```
Comment (Paulami30) [inline docs/tables/aws_waf_rate_based_rule.md:19]:
```suggestion
### List rate-based rules that allows or blocks a request based on the negation of the settings in the predicates
```

**Review by Paulami30 — COMMENTED** [review]
Please check for the comments, run the integration test and update the log.

File: aws/table_aws_waf_rate_based_rule.go (hunk 5 above)
Comment (Priyanka585464) [reply aws/table_aws_waf_rate_based_rule.go:152]: Yes

File: aws-test/tests/aws_waf_rate_based_rule/variables.tf [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+  provider = aws.alternate
+}
+
+data "null_data_source" "resource" {
+  inputs = {
+    scope = "arn:${data.aws_partition.current.partition}:::${data.aws_caller_identity.current.account_id}"
+  }
+}
+
+resource "aws_waf_ipset" "ipset" {
```
Comment (khushboo9024) [inline aws-test/tests/aws_waf_rate_based_rule/variables.tf:49]: Why we need `aws_waf_ipset` ?

File: aws/service.go [hunk 14]
```diff
@@ -1155,6 +1156,22 @@ func StsService(ctx context.Context, d *plugin.QueryData) (*sts.STS, error) {
 	return svc, nil
 }
 
+// WafService returns the service connection for AWS WAF service
+func WafService(ctx context.Context, d *plugin.QueryData) (*waf.WAF, error) {
```
Comment (khushboo9024) [inline aws/service.go:1160]:
```suggestion
func WAFService(ctx context.Context, d *plugin.QueryData) (*waf.WAF, error) {
```
This function is not correct, there is no return and bracket is also missing, this is not going to work.

File: aws/table_aws_waf_rate_based_rule.go (hunk 5 above)
Comment (khushboo9024) [reply aws/table_aws_waf_rate_based_rule.go:152]: No , these are not transform function.

File: aws/table_aws_waf_rate_based_rule.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+	c, err := getCommonColumns(ctx, d, h)
+	if err != nil {
+		return nil, err
+	}
+	commonColumnData := c.(*awsCommonColumnData)
+	aka := "arn:" + commonColumnData.Partition + ":waf::" + commonColumnData.AccountId + ":ratebasedrule" + "/" + id
+	return []string{aka}, nil
+}
+
+func tagListToTurbotTags(ctx context.Context, d *transform.TransformData) (interface{}, error) {
```
Comment (khushboo9024) [inline aws/table_aws_waf_rate_based_rule.go:198]:
```suggestion
func wafRateBasedRuletagListToTurbotTags(ctx context.Context, d *transform.TransformData) (interface{}, error) {
```

**Review by khushboo9024 — CHANGES_REQUESTED** [review]
Minor changes required and please run the integration test and update logs for the same.

File: aws/table_aws_waf_rate_based_rule.go [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+			{
+				Name:        "rate_key",
+				Description: "The field that AWS WAF uses to determine if requests are likely arriving from single source and thus subject to rate monitoring.",
+				Type:        proto.ColumnType_STRING,
+				Hydrate:     getAwsWafRateBasedRule,
+			},
+			{
+				Name:        "rate_limit",
+				Description: "The maximum number of requests, which have an identical value in the field specified by the RateKey, allowed in a five-minute period.",
+				Type:        proto.ColumnType_STRING,
```
Comment (Subhajit97) [inline aws/table_aws_waf_rate_based_rule.go:51]: Seems like the value is of type integer?
Can we convert the value to `int`, or it stores the `string` value?

File: aws/table_aws_waf_rate_based_rule.go [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+			},
+			{
+				Name:        "tags_src",
+				Description: "A list of tags assigned to the Rule.",
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     listAwsWafRateBasedRuleTags,
+				Transform:   transform.FromField("TagInfoForResource.TagList"),
+			},
+
+			// Standard columns for all tables
```
Comment (Subhajit97) [inline aws/table_aws_waf_rate_based_rule.go:69]:
```suggestion
			// Steampipe standard columns
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getAwsWafRateBasedRuleAkas,
+				Transform:   transform.FromValue(),
+			},
+		}),
+	}
+}
+
+//// LIST FUNCTION
+func listAwsWafRateBasedRules(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```
Comment (Subhajit97) [inline aws/table_aws_waf_rate_based_rule.go:95]: Add a new line
```suggestion

func listAwsWafRateBasedRules(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+	plugin.Logger(ctx).Trace("listAwsWafRateBasedRules")
+	// Create session
+	svc, err := WAFService(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+	// List call
```
Comment (Subhajit97) [inline aws/table_aws_waf_rate_based_rule.go:102]: Add new line before comments
```suggestion

	// List call
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+		return nil, err
+	}
+	// Build the params
+	params := &waf.GetRateBasedRuleInput{
+		RuleId: &id,
+	}
+	// Get call
+	data, err := svc.GetRateBasedRule(params)
+	if err != nil {
+		logger.Debug("getAwsWafRateBasedRules", "ERROR", err)
```
Comment (Subhajit97) [inline aws/table_aws_waf_rate_based_rule.go:145]:
```suggestion
		logger.Debug("getAwsWafRateBasedRule", "ERROR", err)
```

File: aws/table_aws_waf_rate_based_rule.go [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+	if err != nil {
+		return nil, err
+	}
+	commonColumnData := commonAwsColumns.(*awsCommonColumnData)
+	// Create Session
+	svc, err := WAFService(ctx, d)
+	if err != nil {
+		return nil, err
+	}
+	aka := "arn:" + commonColumnData.Partition + ":waf::" + commonColumnData.AccountId + ":ratebasedrule" + "/" + id
```
Comment (Subhajit97) [inline aws/table_aws_waf_rate_based_rule.go:173]:
```suggestion
	aka := "arn:" + commonColumnData.Partition + ":waf::" + commonColumnData.AccountId + ":ratebasedrule/" + id
```

**Comment by anisadas** [comment]
```
SETUP: tests/aws_waf_rate_based_rule []

PRETEST: tests/aws_waf_rate_based_rule

TEST: tests/aws_waf_rate_based_rule
Running terraform
data.aws_caller_identity.current: Refreshing state...
data.aws_partition.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.aws_region.alternate: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_waf_rate_based_rule.named_test_resource: Creating...
aws_waf_rate_based_rule.named_test_resource: Creation complete after 8s [id=6178adfb-f3b2-406f-b0d0-859a975d1a34]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

account_id = 986325076436
aws_partition = aws
resource_aka = arn:aws:waf::986325076436:ratebasedrule/6178adfb-f3b2-406f-b0d0-859a975d1a34
resource_id = 6178adfb-f3b2-406f-b0d0-859a975d1a34
resource_name = turbottest63437

Running SQL query: test-get-query.sql
[
  {
    "metric_name": "turbottest63437",
    "partition": "aws",
    "title": "turbottest63437"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:waf::986325076436:ratebasedrule/6178adfb-f3b2-406f-b0d0-859a975d1a34"
    ],
    "metric_name": "turbottest63437",
    "tags": {
      "name": "turbottest63437"
    },
    "title": "turbottest63437"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "metric_name": "turbottest63437",
    "partition": "aws",
    "title": "turbottest63437"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "account_id": "986325076436",
    "akas": [
      "arn:aws:waf::986325076436:ratebasedrule/6178adfb-f3b2-406f-b0d0-859a975d1a34"
    ],
    "title": "turbottest63437"
  }
]
✔ PASSED

POSTTEST: tests/aws_waf_rate_based_rule

TEARDOWN: tests/aws_waf_rate_based_rule

SUMMARY:

1/1 passed.
```

**Review by Subhajit97 — APPROVED** [review]
LGTM

**Review by khushboo9024 — APPROVED** [review]

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**